sklearn/utils/validation.py

```diff
@@ -1100,7 +1100,11 @@ def is_sparse(dtype):
     # See https://pandas.pydata.org/docs/dev/user_guide/copy_on_write.html#read-only-numpy-arrays
     # for more details about pandas copy-on-write mechanism, that is enabled by
     # default in pandas 3.0.0.dev.
-    if _is_pandas_df_or_series(array_orig) and hasattr(array, "flags"):
+    if (
+        _is_pandas_df_or_series(array_orig)
+        and hasattr(array, "flags")
+        and not array.flags.writeable
```

Comment:
If the array is read from a read only buffer such that its writeable is expected to be False as we discussed, won't this be an issue?

Reply:
It is and I opened https://github.com/scikit-learn/scikit-learn/pull/29018 to treat this case correctly for arrays and dataframes. See in particular this https://github.com/scikit-learn/scikit-learn/pull/29018#issuecomment-2127171954 where I show that our current code fails for a very simple use case.

Reply:
Can we try this or will this result in other issues or have unnecessary memory problems.
```python
try:
    array.flags.writeable = True
except ValueError:
    array = array.copy()
    array.flags.writeable = True
```

Reply:
I ruled this out because I think that we should not change the writeability of the user's input data to prevent catastrophic mistakes. Let's discuss this kind of questions in #29018.

Reply:
Okay then can we add the test with as little of the original reproducer as possible? Currently the change looks a bit confusing generally.